server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/remote/delivery/DeliveryRunnable.java

```diff
@@ -52,77 +55,65 @@
     private final MetricFactory metricFactory;
     private final Bouncer bouncer;
     private final MailDelivrer mailDelivrer;
-    private final AtomicBoolean isDestroyed;
     private final Supplier<Date> dateSupplier;
+    private Disposable disposable;
 
     public DeliveryRunnable(MailQueue queue, RemoteDeliveryConfiguration configuration, DNSService dnsServer, MetricFactory metricFactory,
-                            MailetContext mailetContext, Bouncer bouncer, AtomicBoolean isDestroyed) {
+                            MailetContext mailetContext, Bouncer bouncer) {
         this(queue, configuration, metricFactory, bouncer,
             new MailDelivrer(configuration, new MailDelivrerToHost(configuration, mailetContext), dnsServer, bouncer),
-            isDestroyed, CURRENT_DATE_SUPPLIER);
+            CURRENT_DATE_SUPPLIER);
     }
 
     @VisibleForTesting
     DeliveryRunnable(MailQueue queue, RemoteDeliveryConfiguration configuration, MetricFactory metricFactory, Bouncer bouncer,
-                     MailDelivrer mailDelivrer, AtomicBoolean isDestroyeds, Supplier<Date> dateSupplier) {
+                     MailDelivrer mailDelivrer, Supplier<Date> dateSupplier) {
         this.queue = queue;
         this.configuration = configuration;
         this.outgoingMailsMetric = metricFactory.generate(OUTGOING_MAILS);
         this.bouncer = bouncer;
         this.mailDelivrer = mailDelivrer;
-        this.isDestroyed = isDestroyeds;
         this.dateSupplier = dateSupplier;
         this.metricFactory = metricFactory;
     }
 
-    @Override
-    public void run() {
+    public void start() {
+        disposable = queue.deQueue()
+            .publishOn(Schedulers.newParallel("RemoteDelivery", configuration.getWorkersThreadCount()))
+            .flatMap(this::runStep)
+            .onErrorContinue(((throwable, nothing) -> LOGGER.error("Exception caught in RemoteDelivery", throwable)))
+            .subscribeOn(Schedulers.elastic())
+            .subscribe();
+    }
+
+    private Mono<Void> runStep(MailQueue.MailQueueItem queueItem) {
+        TimeMetric timeMetric = metricFactory.timer(REMOTE_DELIVERY_TRIAL);
         try {
-            while (!Thread.interrupted() && !isDestroyed.get()) {
-                runStep();
-            }
+            return processMail(queueItem);
+        } catch (Throwable e) {
```

Comment:
Idem

Reply:
How do you ensure rollback without that kind of catch ?

Reply:
`Mono.{doOnError,onErrorResume}`

Reply:
I tried but it makes the code more complex without any gain